src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts

```diff
@@ -212,21 +231,101 @@ export class MarkdownTextTransformer {
 	 * @param editorControl code editor widget
 	 * @param noSelection controls whether there was no previous selection in the editor
 	 */
-	private setEndSelection(endRange: IRange, type: MarkdownButtonType, editorControl: CodeEditorWidget, noSelection: boolean): void {
-		if (!endRange || !editorControl) {
+	private setEndSelection(endRange: IRange, type: MarkdownButtonType, editorControl: CodeEditorWidget, noSelection: boolean, isUndo: boolean): void {
+		if (!endRange || !editorControl || isUndo) {
 			return;
 		}
 		let offset = this.getColumnOffsetForSelection(type, noSelection);
 		if (offset > -1) {
 			let newRange: IRange = {
 				startColumn: endRange.startColumn + offset,
-				endColumn: endRange.endColumn + offset,
 				startLineNumber: endRange.startLineNumber,
+				endColumn: endRange.startColumn + offset,
 				endLineNumber: endRange.endLineNumber
 			};
 			editorControl.setSelection(newRange);
+		} else {
+			if (this.getMarkdownLineType(type) === MarkdownLineType.BEGIN_AND_END_LINES) {
+				let currentSelection = editorControl.getSelection();
+				editorControl.setSelection({
+					startColumn: currentSelection.startColumn + this.getStartTextToInsert(type).length,
+					startLineNumber: currentSelection.startLineNumber,
+					endColumn: currentSelection.endColumn - this.getEndTextToInsert(type).length,
+					endLineNumber: currentSelection.endLineNumber
+				});
+			}
 		}
 	}
+
+	/**
+	 * Determine if user wants to perform an undo operation
+	 * @param selection current user selection
+	 * @param type markdown button type
+	 * @param lineType markdown line type
+	 * @param editorModel text model for the cell
+	 */
+	private isUndoOperation(selection: Selection, type: MarkdownButtonType, lineType: MarkdownLineType, editorModel: TextModel): boolean {
+		if (lineType === MarkdownLineType.BEGIN_AND_END_LINES) {
+			let selectedText = this.getExtendedSelectedText(selection, type, lineType, editorModel);
+			return selectedText && selectedText.startsWith(this.getStartTextToInsert(type)) && selectedText.endsWith(this.getEndTextToInsert(type));
+		} else {
+			return this.everyLineMatchesBeginString(selection, type, editorModel);
+		}
+	}
+
+	/**
+	 * Gets the extended selected text (current selection + potential beginning + ending transformed text)
+	 * @param selection Current selection in editor
+	 * @param type Markdown Button Type
+	 * @param lineType Markdown Line Type
+	 * @param editorModel TextModel
+	 */
+	private getExtendedSelectedText(selection: Selection, type: MarkdownButtonType, lineType: MarkdownLineType, editorModel: TextModel): string {
+		if (lineType === MarkdownLineType.BEGIN_AND_END_LINES) {
+			return editorModel.getValueInRange({
+				startColumn: selection.startColumn - this.getStartTextToInsert(type).length,
```

Comment:
Make sure you're testing where this would overflow at the beginning and end (i.e. select a word that's the only text on the line and doesn't have the pre/post fix characters)

Reply:
Yes, I did. It's handled gracefully in the editor model; no exceptions are thrown.